modules/consensus/consensusset.go

```diff
@@ -222,6 +228,12 @@ func (cs *ConsensusSet) CurrentBlock() (block types.Block) {
 	return block
 }
 
+// Flush will block until the consensus set has finished all in-progress
+// routines.
+func (cs *ConsensusSet) Flush() error {
+	return cs.tg.Flush()
+}
```

Comment:
`cs.CurrentBlock` only needs a ReadLock, which means that it won't wait until `ProcessConsensusChange` has been called on all the modules. That's the problem I was having with the ecosystem test. You don't know when all the modules have finished receiving a block.

Currently, `ProcessConsensusChange` is called on the modules, blocks until they finish processing, and it calls on the modules in the order that they subscribed in. You can use those implementation details to create the strategy above, where you become the last subscriber and then block until the last subscriber has received the block, but then you force the implementation detail that `ProcessConsensusChange` must be blocking at it must call the modules in a knowable order.

I would very much like to go back to asynchronous calls to `ProcessConsensusChange`. It prevents the consensus set from being blocked for potentially long periods of time, and allows for more parallelism in a daemon that currently tends to peg one core. (our bottleneck would still be boltdb though, and I'm not sure increased parallelism would do that much to help it, even on SSDs. We could test it).

Consensus set aside, it's also important for knowing the progress of other modules. For example, the gateway. The consensus set runs, calls `Broadcast`, then... how do we know when the broadcast call has finished? We could spin on the peers that were supposed to receive the block. If there's an error in the sending though, you won't know whether the sender never finished sending or whether the receiver somehow didn't process it. With `Flush()` you can at least be certain that the sender thinks they finished sending.

And same thing with the host. Right now we don't have functions in the TransactionPool that make it easy to tell whether a storage proof has been submitted. We'd have to grab the list of transactions (or maybe subscribe), then iterate through them and identify the storage proof. And even then, we're not sure if it's the right storage proof, and if we're not finding it we're not sure whether we aren't waiting long enough or if the host was having some other issue. But if we expose a `Flush()` call we know that the host at least submitted the storage proof (or it thinks that it did). And we don't need to iterate through the transaction pool to verify that the process is done.

I'm pretty strongly convinced that `Flush()` will make a lot of our testing difficulties go away. It'll be much easier to set up tests that involve large numbers of asynchronous moving parts - something that we very much struggle with today.